First iteration of adding water tracer infrastructure

billsacks opened the pull request:
This PR adds `shr_wtracers_mod` to coordinate adding arbitrary water tracers across CESM. This also adds `shr_string_withoutSuffix`, which can be used to determine if a given field is a water tracer field, and if so, return the name of the field without the water tracer suffix.

I'll be adding more functionality in a follow-up PR, but some of the changes here are needed for a CMEPS PR that I'm about to open.

billsacks committed:
Add shr_wtracers_mod to coordinate water tracers across the system

billsacks committed:
Treat lowercase species names the same as uppercase

billsacks committed:
Add a logical function: are water tracers present in this simulation?

billsacks committed:
Move some unit tests

Move some shr_string unit tests to keep all of the list-related unit
tests together.

billsacks committed:
Add shr_string_withoutSuffix

billsacks committed:
Make out_str optional in shr_string_withoutSuffix

billsacks committed:
Add WTRACERS_SUFFIX

billsacks commented:
@nusbaume and @mvertens - I'm adding both of you as reviewers. I think a review from one of you is sufficient. @nusbaume - I at least want you to be aware of this and look through it at some point, but I don't know how much you're working over the holidays... if you're not working much, then your review can wait until after the holidays... I might merge it in the meantime if I get a review from @mvertens but I'll be happy to make any changes in a follow-up PR.

billsacks committed:
Remove unused variable

billsacks committed:
Cleanup suggested in review

billsacks commented:
Thank you for the approval, @jedwards4b .

I have done some cleanup suggested by @mvertens - in https://github.com/ESCOMP/CESM_share/pull/78/changes/3881bee450567a7caf8561d721e8685dc7daa793

@mvertens - I'll wait to merge this until you give your approval. Note that I decided not to do the pre-setting of rc = ios (or similar), because I felt that the extra information printed would have been unhelpful (and at least as confusing as helpful because there's no context to help you interpret an arbitrary error code integer).

mvertens reviewed (changes requested):
@billsacks - I have been thinking that specifying mainproc at the module level as part of the shr_waters_init would be better than having to pass in as an argument to the routines. That way if you decide you want to add a new log message on mainproc for a routine that does not have mainproc as an argument - you can easily do that. I've taken that approach in my CDEPS refactor.

billsacks committed:
Store maintask logical

This way it can be used later without always needing to pass it as an
argument.

billsacks commented:
> I have been thinking that specifying mainproc at the module level as part of the shr_waters_init would be better than having to pass in as an argument to the routines. That way if you decide you want to add a new log message on mainproc for a routine that does not have mainproc as an argument - you can easily do that. I've taken that approach in my CDEPS refactor.

Great idea - thanks, @mvertens . I have made that change.

billsacks commented:
By the way, I wanted to call your reviewer attention to one aspect of this new module that is somewhat non-intuitive. As I document at the top of the module:

```Fortran
   ! Note that the init routine uses ESMF-style error handling (where a return code is
   ! returned up the call stack), whereas other routines abort directly if an error is
   ! detected. (The rationale for this is that the init routine has a lot of interaction
   ! with ESMF and is expected to be called from an ESMF-centric part of the code, whereas
   ! the other routines are less ESMF-centric in both their implementation and where they
   ! are expected to be called from, and so it doesn't make as much sense for them to
   ! follow the ESMF error handling paradigm).
```

Please let me know if you have any concerns about this asymmetry.

mvertens commented:
@billsacks - now I'm starting to wonder about my suggestion regarding mainproc. This will work fine if you have all the data model components sharing the same root processor. But what if you have two data components (e.g. DATM, DROF) running on mutually exclusive processors - or only sharing a subset of the processors - then mainproc is no longer uniquely defined. The same would go for using a module level logunit.

billsacks commented:
> now I'm starting to wonder about my suggestion regarding mainproc. This will work fine if you have all the data model components sharing the same root processor. But what if you have two data components (e.g. DATM, DROF) running on mutually exclusive processors - or only sharing a subset of the processors - then mainproc is no longer uniquely defined. The same would go for using a module level logunit.

Ah, very good point. I'll back out that most recent commit.

billsacks committed:
Revert "Store maintask logical"

This reverts commit b441fa67e8cbbfc7dad77917393d8b3e7f1168a4.

As Mariana pointed out: "now I'm starting to wonder about my suggestion
regarding mainproc. This will work fine if you have all the data model
components sharing the same root processor. But what if you have two
data components (e.g. DATM, DROF) running on mutually exclusive
processors - or only sharing a subset of the processors - then mainproc
is no longer uniquely defined. The same would go for using a module
level logunit."
